File: 04_implementation/backend/app/schemas/budget.py
from __future__ import annotations

from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field


class BudgetApuCreate(BaseModel):
    codigo_apu: str = Field(..., min_length=1)
    cantidad: Decimal = Decimal("1")


class BudgetItemRead(BaseModel):
    row_number: int
    item: int
    codigo_apu: str
    descripcion_apu: str
    unidad: str
    cantidad: Decimal
    material_formula: Optional[str] = None
    mano_obra_formula: Optional[str] = None
    valor_unitario_formula: Optional[str] = None
    valor_total_formula: Optional[str] = None


class BudgetAddResponse(BaseModel):
    item: BudgetItemRead
    row_number: int
    backup_path: str
    formula_integrity_ok: bool


class BudgetSaveRequest(BaseModel):
    items: list[BudgetApuCreate] = Field(..., min_length=1)
    sheet_name: str = "PRESUPUESTO"


class BudgetSaveResponse(BaseModel):
    sheet_name: str
    rows: int
    backup_path: str
    formula_integrity_ok: bool


class BudgetExportRequest(BudgetSaveRequest):
    export_code: str = Field(..., min_length=1, max_length=80)


class BudgetExportResponse(BaseModel):
    file_name: str
    file_path: str
    sheet_name: str
    rows: int
    apus: int
    materials: int

File: 04_implementation/frontend/src/api/budget.ts
import { request } from "./client";

export type BudgetItem = {
  row_number: number;
  item: number;
  codigo_apu: string;
  descripcion_apu: string;
  unidad: string;
  cantidad: string;
  material_formula: string | null;
  mano_obra_formula: string | null;
  valor_unitario_formula: string | null;
  valor_total_formula: string | null;
};

export type BudgetListResponse = {
  items: BudgetItem[];
  total: number;
};

export type BudgetAddApuResponse = {
  item: BudgetItem;
  row_number: number;
  backup_path: string;
  formula_integrity_ok: boolean;
};

export type BudgetSaveItemPayload = {
  codigo_apu: string;
  cantidad: string;
};

export type BudgetSaveResponse = {
  sheet_name: string;
  rows: number;
  backup_path: string;
  formula_integrity_ok: boolean;
};

export type BudgetExportResponse = {
  file_name: string;
  file_path: string;
  sheet_name: string;
  rows: number;
  apus: number;
  materials: number;
};

export function listBudgetItems(): Promise<BudgetListResponse> {
  return request<BudgetListResponse>("/budget");
}

export function addApuToBudget(
  codigo_apu: string,
  cantidad = "1",
): Promise<BudgetAddApuResponse> {
  return request<BudgetAddApuResponse>("/budget/apus", {
    method: "POST",
    body: JSON.stringify({ codigo_apu, cantidad }),
  });
}

export function saveBudget(
  items: BudgetSaveItemPayload[],
  sheetName = "PRESUPUESTO",
): Promise<BudgetSaveResponse> {
  return request<BudgetSaveResponse>("/budget/save", {
    method: "POST",
    body: JSON.stringify({ items, sheet_name: sheetName }),
  });
}

export function exportFilteredBudget(
  items: BudgetSaveItemPayload[],
  exportCode: string,
  sheetName = "PRESUPUESTO",
): Promise<BudgetExportResponse> {
  return request<BudgetExportResponse>("/budget/export", {
    method: "POST",
    body: JSON.stringify({
      items,
      sheet_name: sheetName,
      export_code: exportCode,
    }),
  });
}
